File: ssba-2022-01/session-06-files/vm-simulation/vm.py
# ANSI control codes for text formatting
GRAY = '\033[90m'
BLUE = '\033[95m'
BOLD = '\033[1m'
ENDC = '\033[0m'

# Instruction set
LOAD_WORD  = 0x01
STORE_WORD = 0x02
ADD        = 0x03
SUB        = 0x04
ADDI       = 0x05
SUBI       = 0x06
JUMP       = 0x07
BEQZ       = 0x08
HALT       = 0xff

# Loads the word (2 bytes) starting at the given address
# in the memory. Assumes little-endian byte order.
def load_word(memory, addr):
    lo = memory[addr]
    hi = memory[addr + 1]
    return 256 * hi + lo

# Stores the word (2 bytes) starting at the given address
# in the memory, in little-endian byte order.
def store_word(memory, addr, word):
    lo = word % 256
    hi = word // 256
    memory[addr] = lo
    memory[addr + 1] = hi

def print_state(registers, memory):
    print('Current State')

    # Registers
    print()
    print(f'    r1: {registers[1]}')
    print(f'    r2: {registers[2]}')
    print()

    pc = registers[0]

    # Next instruction
    print(f'    pc: 0x{pc:02x} -- ', end='')
    if memory[pc] == HALT:
        print('halt')
    arg1 = memory[pc + 1]
    if memory[pc] == JUMP:
        print(f'jump {arg1:02x}')
    arg2 = memory[pc + 2]
    if memory[pc] == LOAD_WORD:
        print(f'load r{arg1} (0x{arg2:02x})')
    elif memory[pc] == STORE_WORD:
        print(f'store r{arg1} (0x{arg2:02x})')
    elif memory[pc] == ADD:
        print(f'add r{arg1} r{arg2}')
    elif memory[pc] == SUB:
        print(f'sub r{arg1} r{arg2}')
    elif memory[pc] == ADDI:
        print(f'addi r{arg1} {arg2}')
    elif memory[pc] == SUBI:
        print(f'subi r{arg1} {arg2}')
    elif memory[pc] == BEQZ:
        print(f'beqz r{arg1} {arg2}')

    # ASCII arrow to pc location
    print('    ' + GRAY + '-' * (3 * pc + 1) + 'v' + ENDC)

    # Memory
    print('    ' + GRAY + '00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17' + ENDC)
    print('    ' + GRAY + '-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --' + ENDC)
    print('    ' + BOLD + ' '.join(map(lambda x: f'{x:02x}', memory)) + ENDC)
    print('    ' + GRAY + 'INSTRUCTIONS ---------------------------------------^ OUT-^ IN-1^ IN-2^' + ENDC)
    print()

# Simulate the virtual machine with the given memory, with the
# program counter starting at 0x00. The machine will continue
# the "fetch-decode-execute" loop until it reaches a "halt"
# instruction, at which point it will terminate.
def run_vm(memory):

    # Initialize register state
    # Note that unlike `memory`, each entry in the register can hold
    # a 16-bit value directly (up to 65535)
    registers = [
      0x0000, # program counter / instruction pointer
      0x0000, # general purpose register 1
      0x0000, # general purpose register 2
    ]

    while True:
        pc = registers[0]
        instruction = memory[pc]

        print_state(registers, memory)
        input()

        # Handle 0-arg instruction (halt) as special case
        if instruction == HALT:
            return

        arg1 = memory[pc + 1]

        # Handle 1-arg instruction (jump) as special case
        if instruction == JUMP:
            registers[0] = arg1
            continue

        arg2 = memory[pc + 2]

        # load_word reg (addr)
        if instruction == LOAD_WORD:
            registers[arg1] = load_word(memory, arg2)

        # store_word reg (addr)
        elif instruction == STORE_WORD:
            store_word(memory, arg2, registers[arg1])

        # add reg1 reg2
        elif instruction == ADD:
            registers[arg1] += registers[arg2]

        # sub reg1 reg2
        elif instruction == SUB:
            registers[arg1] -= registers[arg2]

        # addi reg x
        elif instruction == ADDI:
            registers[arg1] += arg2

        # subi reg x
        elif instruction == SUBI:
            registers[arg1] -= arg2

        # beqz reg offset
        elif instruction == BEQZ:
            if registers[arg1] == 0:
                registers[0] += arg2

        else:
            raise Exception('Unrecognized instruction:', instruction)

        # All 2-arg instructions advance the program counter
        # by the same amount.
        registers[0] += 3

if __name__ == '__main__':
    """
    00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17
    __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __
    INSTRUCTIONS ---------------------------------------^ OUT-^ IN-1^ IN-2^
    """
    memory = [
        0x01, 0x01, 0x14, # load_word r1 (0x14)
        0x08, 0x01, 0x08, # beqz r1 0x08
        0x03, 0x02, 0x01, # add r2 r1
        0x06, 0x01, 0x01, # subi r1 1
        0x07, 0x03,       # jump 3
        0x02, 0x02, 0x12, # store r2 (0x12)
        0xff,             # halt

        0x00, 0x00,       # output

        0x0a, 0x00,       # input 1
        0x00, 0x00        # input 2
    ]

    run_vm(memory)

File: ssba-2022-01/session-06-files/vm-simulation/vm.c
#include <assert.h>
#include <ncurses.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define NOOP   0x00
#define LOADW  0x01
#define STOREW 0x02
#define ADD    0x03
#define SUB    0x04
#define ADDI   0x05
#define SUBI   0x06
#define JUMP   0x07
#define BEQZ   0x08
#define HALT   0xff

// TODO: highlight recent updates to non-PC registers
void print_registers(uint16_t *registers) {
	printw("\tPC: ");

	attron(A_BOLD);
	printw("0x%02x\n", registers[0]);
	attroff(A_BOLD);

	printw("\tR1: ");

	attron(A_BOLD);
	printw("%d\n", registers[1]);
	attroff(A_BOLD);

	printw("\tR2: ");

	attron(A_BOLD);
	printw("%d\n", registers[2]);
	attroff(A_BOLD);

	addch('\n');
}

int get_opsize(uint16_t pc, uint8_t *memory) {
	uint8_t op = memory[pc];
	int opsize = 3;
	if (op == HALT || op == NOOP) {
		opsize = 1;
	} else if (op == JUMP) {
		opsize = 2;
	}
	return opsize;
}

// TODO: highlight recent updates to memory
void print_memory(uint16_t pc, uint8_t *memory) {
	int opsize = get_opsize(pc, memory);

	// ASCII arrow to pc location
	addch('\t');
	for (int i = 0; i < 3 * pc; i++) {
		addch(' ');
	}
	printw("PC\n");
	addch('\t');
	for (int i = 0; i < 3 * pc + 1; i++) {
		addch(' ');
	}
	addch('v');
	addch('\n');

	// Memory
	addch('\t');
	for (int i = 0; i < 24; i++) {
		if (i > 0) {
			addch(' ');
		}
		printw("%02x", i);
	}
	addch('\n');
	addch('\t');
	for (int i = 0; i < 24; i++) {
		if (i > 0) {
			addch(' ');
		}
		addch(ACS_HLINE);
		addch(ACS_HLINE);
	}
	addch('\n');

	attron(A_BOLD);
	addch('\t');
	for (int i = 0; i < 24; i++) {
		if (i == pc) {
			if (i > 0) {
				addch(' ');
			}
			attron(A_STANDOUT);
		} else if (i == pc + opsize) {
			attroff(A_STANDOUT);
			addch(' ');
		} else if (i > 0) {
			addch(' ');
		}
		printw("%02x", memory[i]);
	}
	addch('\n');
	attroff(A_BOLD);

	printw("\tINSTRUCTIONS ");
	for (int i = 0; i < 39; i++) {
		addch(ACS_HLINE);
	}
	printw("^ OUT-^ IN-1^ IN-2^\n");
	addch('\n');
}

void print_instruction(uint16_t pc, uint8_t *memory) {
	uint8_t op = memory[pc];
	if (op == HALT) {
		printw("HALT");
	} else if (op == NOOP) {
		printw("NOOP");
	} else if (op == JUMP) {
		printw("JUMP\t0x%02x", memory[pc + 1]);
	} else {
		uint8_t arg1 = memory[pc + 1];
		uint8_t arg2 = memory[pc + 2];

		switch (op) {
			case LOADW:
				printw("LOADW\tR%d\t(0x%02x)", arg1, arg2);
				break;

			case STOREW:
				printw("STOREW\tR%d\t(0x%02x)", arg1, arg2);
				break;

			case ADD:
				printw("ADD\tR%d\tR%d", arg1, arg2);
				break;

			case SUB:
				printw("SUB\tR%d\tR%d", arg1, arg2);
				break;

			case ADDI:
				printw("ADDI\tR%d\t%d", arg1, arg2);
				break;

			case SUBI:
				printw("SUBI\tR%d\t%d", arg1, arg2);
				break;

			case BEQZ:
				printw("BEQZ\tR%d\t0x%02x", arg1, arg2);
				break;
		}
	}
	addch('\n');
}

void print_instructions(uint16_t pc, uint8_t *memory) {
	for (int i = 0; i < 18; i += get_opsize(i, memory)) {
		addch('\t');
		printw("0x%02x\t", i);

		attron(A_BOLD);
		if (i == pc) {
			attron(A_STANDOUT);
			print_instruction(i, memory);
			attroff(A_STANDOUT);
		} else {
			print_instruction(i, memory);
		}
		attroff(A_BOLD);
	}
}

void title(char *name, int size) {
	attron(A_STANDOUT);
	printw("%s", name);
	for (int i = strlen(name); i < size; i++) {
		addch(' ');
	}
	attroff(A_STANDOUT);
	printw("\n\n");
}

void print_state(uint16_t *registers, uint8_t *memory) {
	clear();
	move(0, 0);

	uint16_t pc = registers[0];

	title("REGISTERS", 80);
	print_registers(registers);

	title("MEMORY", 80);
	print_memory(pc, memory);

	title("PROGRAM", 80);
	print_instructions(pc, memory);

	refresh();
}

void run_vm(uint8_t *memory) {
	uint16_t registers[3] = {
		0x0000,
		0x0000,
		0x0000
	};

	while (1) {
		print_state(registers, memory);
		getch();

		uint16_t pc = registers[0];
		uint8_t instruction = memory[pc];

		// Handle HALT and NOOP separately because they don't have any args.
		if (instruction == HALT) {
			return;
		}

		if (instruction == NOOP) {
			registers[0]++;
			continue;
		}

		// Handle JUMP separately because it only has one arg.
		if (instruction == JUMP) {
			registers[0] = memory[pc + 1];
			continue;
		}

		uint8_t arg1 = memory[pc + 1];
		uint8_t arg2 = memory[pc + 2];

		uint8_t *byte_pointer;
		uint16_t *word_pointer;

		switch (instruction) {
			// load_word reg (addr)
			case LOADW:
				byte_pointer = memory + arg2;
				word_pointer = (uint16_t *)byte_pointer;

				registers[arg1] = *word_pointer;
				break;

			// store_word reg (addr)
			case STOREW:
				byte_pointer = memory + arg2;
				word_pointer = (uint16_t *)byte_pointer;

				*word_pointer = registers[arg1];
				break;

			// add reg1 reg2
			case ADD:
				registers[arg1] += registers[arg2];
				break;

			// sub reg1 reg2
			case SUB:
				registers[arg1] -= registers[arg2];
				break;

			case ADDI:
				registers[arg1] += arg2;
				break;

			case SUBI:
				registers[arg1] -= arg2;
				break;

			case BEQZ:
				if (registers[arg1] == 0) {
					registers[0] += arg2;
				}
				break;

			default:
				endwin();
				fprintf(stderr, "Unrecognized instruction: 0x%x\n", instruction);
				exit(1);
		}

		// All 2-arg instructions (even BEQZ) advance PC by the same amount.
		registers[0] += 3;
	}
}

int main() {
	uint8_t memory[24] = {
		0x01, 0x01, 0x14, // load_word r1 (0x14)
		0x08, 0x01, 0x08, // beqz r1 0x08
		0x03, 0x02, 0x01, // add r2 r1
		0x06, 0x01, 0x01, // subi r1 1
		0x07, 0x03,       // jump 3
		0x02, 0x02, 0x12, // store r2 (0x12)
		0xff,             // halt

		0x00, 0x00,       // output

		0x0a, 0x00,       // input 1
		0x00, 0x00        // input 2
	};

	initscr();
	cbreak();
	curs_set(0);
	noecho();
	run_vm(memory);
	endwin();
}
